File: Sources/MuPDFMobile/MuPDFAnnotation.swift
import Foundation

// MARK: - MuPDFAnnotation

/// Represents a single PDF annotation on a ``MuPDFPage``.
///
/// After modifying any property call ``update()`` to flush the change to the
/// underlying PDF structure.
#if canImport(ObjectiveC)
@objcMembers
#endif
public final class MuPDFAnnotation: NSObject {

    // -------------------------------------------------------------------------
    // MARK: Properties
    // -------------------------------------------------------------------------

    /// The type of this annotation.
    public private(set) var type: MuPDFAnnotationType

    /// Bounding rectangle of the annotation in PDF user-space points.
    public var rect: MuPDFRect {
        didSet { isDirty = true }
    }

    /// The annotation's stroke / fill colour.
    public var color: MuPDFColor {
        didSet { isDirty = true }
    }

    /// Opacity of the annotation in [0, 1]. 1 = fully opaque.
    public var opacity: Float {
        didSet {
            opacity = min(1, max(0, opacity))
            isDirty = true
        }
    }

    /// Text content / comment associated with the annotation.
    public var contents: String {
        didSet { isDirty = true }
    }

    // -------------------------------------------------------------------------
    // MARK: Private state
    // -------------------------------------------------------------------------

    private var nativeAnnot: OpaquePointer?
    private weak var page: MuPDFPage?
    private var isDirty = false

    // -------------------------------------------------------------------------
    // MARK: Init
    // -------------------------------------------------------------------------

    init(
        type: MuPDFAnnotationType,
        rect: MuPDFRect,
        page: MuPDFPage,
        nativeAnnot: OpaquePointer? = nil
    ) {
        self.type       = type
        self.rect       = rect
        self.color      = .highlightYellow
        self.opacity    = 1.0
        self.contents   = ""
        self.page       = page
        self.nativeAnnot = nativeAnnot
    }

    deinit {
        // TODO: fz_drop_annot(MuPDFContext.shared.ctx, nativeAnnot)
    }

    // -------------------------------------------------------------------------
    // MARK: Update
    // -------------------------------------------------------------------------

    /// Flushes any pending property changes to the underlying PDF annotation.
    ///
    /// Call this after modifying ``rect``, ``color``, ``opacity``, or
    /// ``contents``. Changes are not persisted to disk until
    /// ``MuPDFDocument/save(to:)`` is called.
    public func update() {
        guard isDirty else { return }
        isDirty = false

        // TODO: (requires mupdf submodule)
        //   let ctx = MuPDFContext.shared.ctx
        //   // Update rect
        //   let (x0, y0, x1, y1) = toFzRect(rect)
        //   pdf_set_annot_rect(ctx, nativeAnnot, fz_make_rect(x0, y0, x1, y1))
        //   // Update colour
        //   var colorComponents = toFzColor(color)
        //   pdf_set_annot_color(ctx, nativeAnnot, Int32(colorComponents.count), &colorComponents)
        //   // Update opacity
        //   pdf_set_annot_opacity(ctx, nativeAnnot, opacity)
        //   // Update contents
        //   pdf_set_annot_contents(ctx, nativeAnnot, contents)
        //   pdf_update_annot(ctx, nativeAnnot)
    }

    // -------------------------------------------------------------------------
    // MARK: Description
    // -------------------------------------------------------------------------

    public override var description: String {
        "MuPDFAnnotation(type: \(type.name), rect: \(rect), opacity: \(opacity))"
    }
}

File: android/lib/src/main/kotlin/com/artifex/mupdf/mobile/MuPDFAnnotation.kt
package com.artifex.mupdf.mobile

/**
 * Represents a single PDF annotation on a [MuPDFPage].
 *
 * Modifications to properties are staged in memory. Call [update] to flush
 * them to the underlying PDF structure. Changes are not written to disk until
 * [MuPDFDocument.save] is called.
 *
 * @property type     The type of this annotation.
 * @property rect     Bounding rectangle in PDF user-space points.
 * @property color    Stroke / fill colour of the annotation.
 * @property opacity  Opacity in [0, 1]. 1 = fully opaque.
 * @property contents Text content / comment associated with the annotation.
 */
open class MuPDFAnnotation internal constructor(
    val type: MuPDFAnnotationType,
    var rect: MuPDFRect,
    var color: MuPDFColor = MuPDFColor.HighlightYellow,
    var opacity: Float = 1f,
    var contents: String = "",
    /** Opaque handle to the native `pdf_annot *`. -1 means not yet backed by native. */
    internal val nativeHandle: Long = -1L,
    internal val page: MuPDFPage? = null
) {
    init {
        require(opacity in 0f..1f) { "opacity must be in [0, 1]" }
    }

    /**
     * Flushes any pending property changes to the underlying PDF annotation.
     *
     * Must be called after modifying [rect], [color], [opacity], or
     * [contents] to ensure the changes are reflected when the document is
     * saved.
     */
    open fun update() {
        if (nativeHandle == -1L) return
        // Delegate to the JNI bridge.
        // TODO: (requires mupdf submodule)
        // nativeUpdateAnnotation(
        //     nativeHandle,
        //     rect.x, rect.y, rect.x + rect.width, rect.y + rect.height,
        //     color.r, color.g, color.b, color.a,
        //     opacity,
        //     contents
        // )
    }

    override fun toString(): String =
        "MuPDFAnnotation(type=${type.displayName}, rect=$rect, opacity=$opacity)"

    companion object {
        // JNI bridge — called from update()
        @JvmStatic
        private external fun nativeUpdateAnnotation(
            nativeHandle: Long,
            x0: Float, y0: Float, x1: Float, y1: Float,
            r: Float, g: Float, b: Float, a: Float,
            opacity: Float,
            contents: String
        )
    }
}
